False positive W0231 when using typing.overload.

### Steps to reproduce
Minimum failing example.
```python
from typing import Union, overload


class Parent:

    def __init__(self, num: int) -> None:
        self.number = num


class Child(Parent):

    @overload
    def __init__(self, num: int) -> None:
        ...

    @overload
    def __init__(self, num: float) -> None:
        ...

    def __init__(self, num: Union[int, float]):
        super().__init__(round(num))

```
Run pylint on the above example code.

### Current behavior
```
pylint_init_bug.py:13:4: W0231: __init__ method from base class 'Parent' is not called (super-init-not-called)
pylint_init_bug.py:17:4: W0231: __init__ method from base class 'Parent' is not called (super-init-not-called)
```

### Expected behavior
The methods decorated with `typing.overload` should not be required to call the base `__init__` as they are "for the benefit of the type checker only" as indicated at [https://docs.python.org/3/library/typing.html#typing.overload](https://docs.python.org/3/library/typing.html#typing.overload).



### pylint --version output
```
pylint 2.3.1
astroid 2.2.5
Python 3.7.3 (default, Apr  3 2019, 19:16:38) 
[GCC 8.0.1 20180414 (experimental) [trunk revision 259383]]
```

Same bug exists on the master branch which includes #2927.
```
pylint 2.4.0-dev0
astroid 2.3.0
Python 3.7.3 (default, Apr  3 2019, 19:16:38) 
[GCC 8.0.1 20180414 (experimental) [trunk revision 259383]]
```

--- a/pylint/checkers/classes.py
+++ b/pylint/checkers/classes.py
@@ -1550,7 +1550,7 @@ a metaclass class method.",
 
     def _check_init(self, node):
         """check that the __init__ method call super or ancestors'__init__
-        method
+        method (unless it is used for type hinting with `typing.overload`)
         """
         if not self.linter.is_message_enabled(
             "super-init-not-called"
@@ -1600,6 +1600,8 @@ a metaclass class method.",
             except astroid.InferenceError:
                 continue
         for klass, method in not_called_yet.items():
+            if decorated_with(node, ["typing.overload"]):
+                continue
             cls = node_frame_class(method)
             if klass.name == "object" or (cls and cls.name == "object"):
                 continue

--- a/tests/functional/init_not_called.py
+++ b/tests/functional/init_not_called.py
@@ -62,3 +62,25 @@ from missing import Missing
 
 class UnknownBases(Missing):
     """Don't emit no-init if the bases aren't known."""
+
+
+from typing import overload  # pylint: disable=wrong-import-order
+
+class Parent:
+
+    def __init__(self, num: int):
+        self.number = num
+
+
+class Child(Parent):
+
+    @overload
+    def __init__(self, num: int):
+        ...
+
+    @overload
+    def __init__(self, num: float):
+        ...
+
+    def __init__(self, num):
+        super().__init__(round(num))

